### PR TITLE
Create CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by sanchaez**
`CONTRIBUTING.md`  is a common way to tell external developers on how to contribute. This way we can link to this file in case of git author name issues for example.

### CODE REVIEW

**Comment by HBelusca** [comment]
Since it's a .md file, this means (I guess) that it's mainly for viewing from e.g. github with formatting. Could you confirm?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# How To Contribute
+There are several ways to contribute to the development of ReactOS. The most often encountered problem is not knowing where to begin or what to do. If you are able to program or understand the technical information that is pertinent to this project, helping the development can be easy.
+
+## Documentation
+
+There are some important points if you'd like to help document ReactOS:
```
Comment (HBelusca) [inline CONTRIBUTING.md]: In written texts, it is always better not to use abbreviated forms. So:
you'd --> you would
doesn't --> does not
etc... (there are others in the text),

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+1.  Make sure the documentation doesn't exist yet (if it does, help improve it).
+2.  Respect [clean room reverse engineering] practices.
+3.  Add your knowledge to a place where the other developers can find it.
+
+## Test ReactOS
+
+### Localize bugs
+
+By localizing bugs, developers can identify what causes the bug and what it affects. There are a variety of methods to [debug] ReactOS while testing it. After identifying a bug, check if it is already known about by searching [JIRA] and adding any additional information to the report. If you think that it is an unidentified bug, consider [filing a bug report].
```
Comment (HBelusca) [inline CONTRIBUTING.md]: "what it affects." --> "which parts it affects." (suggestion)

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+### Fix bugs
+
+Instead of looking for bugs, you can also try to fix a few that are already listed in JIRA. Fixing bugs requires a lot more skill than simply searching for them, and can be time consuming.
```
Comment (HBelusca) [inline CONTRIBUTING.md]: "a lot more skill" --> "a lot more skills" (terminating 's') : @gedmurphy or @jimtabor, could you confirm?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+### Write tests
+
+Tests are used to check the functionality and correctness of APIs on ReactOS compared to Windows implementations. There are also some unit tests that you could help ReactOS pass, which can be found [here][testman].
+
+## Implement new things
+
+Considering ReactOS is alpha quality software, there is a lot of [missing functionality] that Windows operating systems have. Before starting a project to implement something, find out if another person is working on the same thing. If you find that someone is already working on it, ask if any assistance is needed for what specifically is being worked on or a related project. Plenty of times a person will start to implement something and never finish before moving to something else. Make sure you stay committed to what you are going to implement, and do not be afraid to ask for assistance if you need help with something.
```
Comment (HBelusca) [inline CONTRIBUTING.md]: "find out if another person" <--- s/if/whether/

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+Since our [migration to GitHub] we gladly accept [Pull Requests]. Pull requests let you tell others about changes you've pushed to a repository on GitHub. Once a pull request is opened, you can discuss and review the potential changes with collaborators and add follow-up commits before the changes are merged into the repository. Pull request is a prefered way to submit your work - it makes reviewing and merging your contribution much easier.
+
+### Patches
+
+A [patch] is a set of changes to existing source code. The changes in a patch can be merged into existing source code. This process is referred to as applying a patch (to source code). Which changes a patch contains and the way the patch is structured can have significant impact on the consequences that can happen from applying the patch. 
+
+See [Submitting Patches] for details.
+
+### Rules and Recomendations
```
Comment (HBelusca) [inline CONTRIBUTING.md]: s/Recomendations/Recommendations/ (with two 'm')

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+Regardless of your way of contributing to ReactOS we have some rules and recomendations:
```
Comment (HBelusca) [inline CONTRIBUTING.md]: ditto

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+- Use your __full real name__ and __real email__. We don't accept anonymous contributions!
+- Ensure your contribution is properly described. Include the relevant issue number if applicable.
+- Put only related changes. It will make reviewing easier as the reviewer needs to recall less information about the existing source code that is changed.
```
Comment (HBelusca) [inline CONTRIBUTING.md]: "recall" or "remember"?

File: CONTRIBUTING.md (hunk 3 above)
Comment (gedmurphy) [reply CONTRIBUTING.md]: You can either have "a lot more skill" or "many more skills" if you want plurality.
"a lot more skill" flows better in this context.

File: CONTRIBUTING.md (hunk 4 above)
Comment (gedmurphy) [reply CONTRIBUTING.md]: 1) Agreed. "whether another person" sounds better.
2) Agreed.  A better way of structuring that sentence would be : "more often than not, someone will start to implement something and move onto something else before it's complete"

**Comment by binarymaster** [comment]
I think this document also should describe, which parts of the project are synced with upstream repos of other projects, like Wine Staging, BTRFS, UniATA, etc.

So the contributor can check whether the patch should be sent to upstream.

**Comment by sanchaez** [comment]
`Since it's a .md file, I guess this means that it's mainly for viewing from e.g. github with formatting.`
Yes.

File: CONTRIBUTING.md (hunk 1 above)
Comment (sanchaez) [reply CONTRIBUTING.md]: FYI this was copied from the Wiki (most of the text).

File: CONTRIBUTING.md (hunk 7 above)
Comment (sanchaez) [reply CONTRIBUTING.md]: `recall` is better in this context IMO. You open the changed code to _recall_ details, but you don't need to _remember_ it before you review.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+## Contribute
+
+Your contribution can be of numerous forms. We currently accept two ways to contribute - Pull Requests and Patches.
+
+### Pull Requests
+
+Since our [migration to GitHub] we gladly accept [Pull Requests]. Pull requests let you tell others about changes you've pushed to a repository on GitHub. Once a pull request is opened, you can discuss and review the potential changes with collaborators and add follow-up commits before the changes are merged into the repository. Pull request is a prefered way to submit your work - it makes reviewing and merging your contribution much easier.
```
Comment (sanchaez) [inline CONTRIBUTING.md]: Maybe putting an emphasis on `Pull request is a prefered way to submit your work` or remove patches section entirely would be doable?

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
+# How To Contribute
+There are several ways to contribute to the development of ReactOS. The most often encountered problem is not knowing where to begin or what to do. If you are able to program or understand the technical information that is pertinent to this project, helping the development can be easy.
+
+## Documentation
+
+There are some important points if you wouldd like to help document ReactOS:
```
Comment (ThFabba) [inline CONTRIBUTING.md]: would*

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+1.  Make sure the documentation does not exist yet (if it does, help improve it).
+2.  Respect [clean room reverse engineering] practices.
+3.  Add your knowledge to a place where the other developers can find it.
```
Comment (ThFabba) [inline CONTRIBUTING.md]: This doesn't seem particularly actionable. Would it make sense to list examples of such places?

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+By localizing bugs, developers can identify what causes the bug and which part it affects. There are a variety of methods to [debug] ReactOS while testing it. After identifying a bug, check if it is already known about by searching on [JIRA] and adding any additional information to the report. If you think that it is an unidentified bug, consider [filing a bug report].
+
+### Fix bugs
+
+Instead of looking for bugs, you can also try to fix a few that are already listed in JIRA. Fixing bugs requires a lot more skill than simply searching for them, and can be time consuming.
+
+### Write tests
+
+Tests are used to check the functionality and correctness of APIs on ReactOS compared to Windows implementations. There are also some unit tests that you could help ReactOS pass, which can be found [here][testman].
```
Comment (ThFabba) [inline CONTRIBUTING.md]: s/here/in the Web Test Manager/
https://www.w3.org/QA/Tips/noClickHere

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+- Respect our Coding Style. See [Coding Style] and [Programming Guidelines] for details on how to write your code.
+- Do not be afraid to ask questions. Ask our developers on JIRA or [IRC] channel.
+
+## Finding a good project to start
+
+Finding a good project to start with can be a challenge, because when starting out you are (usually) not aware of all the possibilities. To help you find a project, here are some ideas to try:
+
+-   Find a test that fails, and try to make it succeed: <https://www.reactos.org/testman/>
+-   Look around in JIRA, and if you have problems finding nice projects to start with, there is a label for this: <https://jira.reactos.org/issues/?jql=labels%20%3D%20starter-project>
+-   Ask for help in [IRC]
```
Comment (ThFabba) [inline CONTRIBUTING.md]: on IRC*

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+  [filing a bug report]: https://reactos.org/wiki/File_Bugs
+  [testman]: http://www.reactos.org/testman/
+  [migration to GitHub]: https://www.reactos.org/project-news/reactos-repository-migrated-github
+  [humans are terrible at tracking large amount of information]: http://www.eurekalert.org/pub_releases/2005-03/aps-hmc030805.php
+  [Pull requests]: https://help.github.com/articles/about-pull-requests/
+  [tips for reviewing patches]: http://drupal.org/patch/review
+  [missing functionality]: https://reactos.org/wiki/Missing_ReactOS_Functionality
+  [patch]: https://git-scm.com/docs/git-format-patch
+  [Submitting Patches]: https://reactos.org/wiki/Submitting_Patches
+  [amend]: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History
```
Comment (ThFabba) [inline CONTRIBUTING.md]: This is not referred to in the text.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+  [Coding Style]: https://reactos.org/wiki/Coding_Style
+  [#reactos]: https://www.reactos.org/irc
```
Comment (ThFabba) [inline CONTRIBUTING.md]: Also not referred to?

**Comment by AmineKhaldi** [comment]
@binarymaster regarding your suggestion we may use ```3rd Party Files.txt```, ```Readme.fsd``` and ```README.WINE``` from the ```media/doc``` folder as references.

**Comment by sanchaez** [comment]
@AmineKhaldi this should be good to merge.

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+You can try to fix a few bugs that are already listed in [JIRA]. Squashing bugs is not a simple task. It requires a lot more skill than simply searching for them, and can be time consuming; however, by doing that you greatly help ReactOS become a stable system.
+
+_NOTE: patches related to 3rd party code such as Wine or BtrFS should be sent to upstream of the said projects. See [3rd Party Files.txt], [README.WINE] and [README.FSD] in [media/doc](media/doc) for details._
+
+### Fix tests
+
+Tests are used to check the functionality and correctness of APIs on ReactOS compared to Windows implementations. There are some unit tests that you could help ReactOS pass, which can be found [in the Web Test Manager][testman] and some that are broken or yet to be written.
+
+### Fixing Coverity scans
```
Comment (AmineKhaldi) [inline CONTRIBUTING.md]: This should be "Fix" for consistency with the rest yeah?

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+
+[Coverity] is enhanced static analysis that uncovers leaks, buffer overflows, security issues and other. We do  such scans  on ReactOS codebase pretty often. You can [request][request-coverity] to see Coverity 'defects' and help to fix them.
```
Comment (AmineKhaldi) [inline CONTRIBUTING.md]: "We do(I see two spaces here)such scans(and here too)on ReactOS"

File: CONTRIBUTING.md (hunk 15 above)
Comment (sanchaez) [reply CONTRIBUTING.md]: Yep, missed that.

**Pull request merged**